catalog/core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/CachingFederationStrategyTest.java

```diff
@@ -0,0 +1,390 @@
+        when(mockResponse.getHits()).thenReturn((long) results.size());
+        when(mockResponse.getResults()).thenReturn(results);
+
+        MockSource mockSource = new MockSource("mock source",
+                "mock title",
+                "mock 1.0",
+                "mock org",
+                new HashSet<>(),
+                true,
+                new Date());
+
+        mockSource.setQueryResponse(mockResponse);
+
+        List<Source> sources = new ArrayList<>();
+        sources.add(mockSource);
+
+        QueryResponse response = strategy.federate(sources, fedQueryRequest);
+
+        WaitCondition.expect("Wait until the result appears in the response")
+                .until(() -> response.getHits(), is(42L))
+                .checkEvery(100, TimeUnit.MILLISECONDS)
+                .within(1, TimeUnit.SECONDS);
+    }
+
+    @Test
+    public void testFederateQueryAndUpdateCacheNotBlocking() throws Exception {
+        HashMap<String, Serializable> properties = new HashMap<>();
+        properties.put(QUERY_MODE, UPDATE_QUERY_MODE);
+
+        QueryRequest fedQueryRequest = new QueryRequestImpl(mockQuery, properties);
+        when(fedQueryRequest.getQuery()
+                .getStartIndex()).thenReturn(2);
+
+        MetacardImpl mockMetacard = new MetacardImpl();
+
+        Result mockResult = new ResultImpl(mockMetacard);
+
+        SourceResponse mockResponse = mock(SourceResponse.class);
+        List<Result> results = Arrays.asList(mockResult);
+        when(mockResponse.getHits()).thenReturn((long) results.size());
+        when(mockResponse.getResults()).thenReturn(results);
+
+        MockSource mockSource = new MockSource("mock source",
+                "mock title",
+                "mock 1.0",
+                "mock org",
+                new HashSet<>(),
+                true,
+                new Date());
+
+        mockSource.setQueryResponse(mockResponse);
+
+        List<Source> sources = new ArrayList<>();
+        sources.add(mockSource);
+
+        QueryResponse response = strategy.federate(sources, fedQueryRequest);
+
+        WaitCondition.expect("Wait until the result appears in the response")
+                .until(() -> response.getHits(), is(42L))
+                .checkEvery(100, TimeUnit.MILLISECONDS)
+                .within(1, TimeUnit.SECONDS);
+    }
+
+    @Test
+    public void testFederateQueryCacheWithOffsetHandler() throws Exception {
+
+        HashMap<String, Serializable> properties = new HashMap<>();
+        properties.put(QUERY_MODE, INDEX_QUERY_MODE);
+
+        Query mockQ = new QueryImpl(mock(NullFilterImpl.class),
+                2,
+                5,
+                mock(SortBy.class),
+                true,
+                1000);
+
+        QueryRequest fedQueryRequest = new QueryRequestImpl(mockQ, properties);
+
+        MetacardImpl mockMetacard = new MetacardImpl();
+
+        Result mockResult = new ResultImpl(mockMetacard);
```

Comment:
If you haven't already, I would recommend investigate how much of this can go in `@Before` to clean up the readability of the unit tests